### ISSUE TITLE
Add `omit_hidden` option to `readdir` ?

### ISSUE THREAD

**Issue by CameronBieganek** [JuliaLang/julia#38841]
I couldn't find an issue for this. I was expecting there to be an optional keyword argument for `readdir` to omit hidden files and directories. Currently I'm doing this

```julia
function non_hidden_files(dir; kwargs...)
    files = readdir(dir; kwargs...)
    filter(!startswith(".") ∘ basename, files)
end
```

which works for hidden files but not for hidden directories. It seems like it would be handy to have an `omit_hidden` or similar option for `readdir`.

I encountered this because of the pesky `.DS_Store` files on MacOS. :joy:

**Comment by StefanKarpinski** [comment]
Define hidden file...

**Comment by CameronBieganek** [comment]
True, I have no idea how hidden files work on Windows. I'm not a Windows user...

**Comment by StefanKarpinski** [comment]
Well, define hidden on macOS then...

**Comment by stevengj** [comment]
On Windows, one could use [GetFileAttributesW](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfileattributesw) and check whether the file has `FILE_ATTRIBUTE_HIDDEN`, I think.

Presumably one could implement something to check for [hidden files on MacOS](https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/FileSystemOverview/FileSystemOverview.html#//apple_ref/doc/uid/TP40010672-CH2-SW7) as well.

How useful would a filesystem-specific `ishidden` function be?

**Comment by StefanKarpinski** [comment]
I think that would be the first step. It could just check for leading dot in the name for Linux; for macOS it would check that plus the three other kinds of hiddenness; for Windows the GetFileAttributesW thing. There are some additional considerations: is a path hidden if some prefix of it is hidden? Or only if the last component is hidden?

**Comment by fonsp** [comment]
This would be great for UI applications written in Julia :) 

It looks like Python 3.5 started including the output of `GetFileAttributesW` in the output of `stat`: https://docs.python.org/3/library/stat.html#stat.FILE_ATTRIBUTE_HIDDEN (It does seem like Python and NodeJS do not have "hidden file" in their stdlib, perhaps because it is not as well-defined outside of windows)

**Comment by jakewilliami** [comment]
Thanks for raising this issue @CameronBieganek .  I agree that this functionality would be useful, although it needn't necessarily be in Base.  As such, I've begun work on a library ([HiddenFiles.jl](https://github.com/jakewilliami/HiddenFiles.jl/)) to hide away all of the nastiness/nuance of hidden files on different systems.  It's very much a work in progress, but it's getting there.  Still need to register the package, but will do that after I make proper tests/CI/documentation, and finish implementing a solution for BSD.

@StefanKarpinski, you mentioned there are three things to check after checking for a leading `.` when it comes to macOS, however, I've only found one other criteria (`UF_HIDDEN`).  Do you have any resources on the others?  Perhaps @stevengj is more familiar with macOS hidden files?

Also, is there a consensus on your earlier question?:
> is a path hidden if some prefix of it is hidden? Or only if the last component is hidden?

**Comment by stevengj** [comment]
See [this link](https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/FileSystemOverview/FileSystemOverview.html#//apple_ref/doc/uid/TP40010672-CH2-SW7) for macOS hidden files.

**Comment by jakewilliami** [comment]
Thanks @stevengj, I saw that link from your previous comment, however it felt very high-level.  If that's the best that there is then I can probably work from that, thank you.
